### ISSUE TITLE
Draw multiple polygons with a single PolygonVisual instance

### ISSUE THREAD

**Issue by sebae1** [vispy/vispy#2228]
Hi.

I am making a drawing program like many usual Vispy users. And faced a lagging trouble like the others.

Originally my code generates individual lines, markers, and polygons, but now modifying them to a single line, marker, and poylgon. MarkerVisual naturally does not have connectivity and LineVisual offers 'connect' argument thus they are not problems. But how should I deal with PolygonVisual?

There is PolygonCollection but it seems that it is unable to be fit with scene yet which I must use.

In summary, one of following would solve my problem.

 - Multiple polygons are drawn by a single PolygonVisual (like 'connect' argument in LineVisual)
 - Parent scene (view box) is specified in PolygonCollection
 - Any other option that does not cause severe slowing even over 100 polygons

Thanks for your help.

**Comment by djhoese** [comment]
Any ideas on this @brisvag or @sofroniewn or @almarklein? My usual response is that this currently isn't possible with polygons. As you've pointed out @sebae1, lines and markers aren't really a problem which is usually what a lot of people are using when they run into these issues. As far as I know we don't currently have a performance-friendly way of making a lot of polygons.

**Comment by brisvag** [comment]
I'm not sure I understand the problem; I never used polygons, so I might be way off, but:

- if you're trying to "connect" polygons just like you connect lines, that's just a generic `MeshVisual`.
- otherwise, if you *don't* want to connect them, the collection seems the only way right now.
 
I *think* that for a performance-friendly approach we can do it by updating `MeshVisual` to accept an arbitrary index buffer, and then make a subclass of that that does some magic behind the scenes to get from polygon definitions to triangle meshes.

**Comment by brisvag** [comment]
@sebae1 can you provide a simple code example of what you're trying to achieve?

**Comment by sebae1** [comment]
> @sebae1 can you provide a simple code example of what you're trying to achieve?

Yep sure.
My program uses Wxpython as a backend. This example code also involves is to reproduce the environment.

DESCRIPTION
- Key event "Q" : camera goes to the default region
- Key event "A" : generates 100 random markers
- Key event "Z" : removes   100 random markers
- Left canvas     : uses **a single MarkerVisual**
- Right canvas  : uses individual MarkerVisuals

If you generate few hundred markers on the right, moving the camera or resizing the frame causes lagging, while the left one does not at all.

Likewise, I want to make a lot of polygons without lagging but can't find the way.

```
import wx
import numpy as np
import random

from vispy import scene

class CustomMarker:

    def __init__(self, parent_scene):
        self.__marker = scene.Markers(parent=parent_scene)
        self.__default_color = (1, 1, 1, 1)
        self.__params = []
        #    pos       color    visible
        # [[x, y], [r, g, b, a], bool]
    
    @property
    def params(self) -> list:
        return self.__params

    def __isOverlapped(self, pos:list):
        current_pos = [",".join(map(str, param[0])) for param in self.__params]
        if ",".join(map(str, pos)) in current_pos:
            return True
        else:
            return False

    def __updateMarkers(self):
        if len(self.__params) == 0:
            self.__marker.set_data(pos=None)
            return
        pos   = np.array([param[0] for param in self.__params if param[2]])
        color = np.array([param[1] for param in self.__params if param[2]])
        self.__marker.set_data(pos=pos, face_color=color)

    def __getMarkerIndex(self, pos:list):
        current_pos = [",".join(map(str, param[0])) for param in self.__params]
        pos = ",".join(map(str, pos))
        if pos not in current_pos:
            return None
        else:
            return current_pos.index(pos)

    def addMarker(self, pos:list, color:list=None, visible:bool=True):
        if color is None: color = self.__default_color
        if self.__isOverlapped(pos):
           return
        self.__params.append([pos, color, visible])
        self.__updateMarkers()
    
    def removeMarker(self, pos:list):
        i = self.__getMarkerIndex(pos)
        if i is None:
            return
        self.__params.pop(i)
        self.__updateMarkers()

class MySceneCanvas1(scene.SceneCanvas):

    def __init__(self, parent):
        scene.SceneCanvas.__init__(self, parent=parent, app="wx", size=(10, 10))

        self.unfreeze()

        grid = self.central_widget.add_grid(margin=0)
        self.__view = grid.add_view(row=0, col=0)
        self.__view.camera = "panzoom"
        self.__view.camera.interactive = True
        self.__view.camera.rect = (0, 0, 1, 1)
        self.__marker = CustomMarker(self.__view.scene)

        self.__info = scene.Text(
            parent   = self.scene,
            pos      = (10, 10),
            anchor_x = "left",
            anchor_y = "bottom",
            color    = (1, 0, 0, 1),
            text     = "0"
        )
        
        self.freeze()

        self.show()

    def on_key_press(self, event):
        N = 100
        if event.text == "Q":
            self.__view.camera.rect = (0, 0, 1, 1)
        elif event.text == "A":
            for i in range(N):
                random_pos = np.random.rand(1, 2)[0]
                self.__marker.addMarker(random_pos, np.random.rand(1, 4)[0])
        elif event.text == "Z":
            for i in range(N):
                if len(self.__marker.params) == 0:
                    break
                random_pos = random.choice([param[0] for param in self.__marker.params])
                self.__marker.removeMarker(random_pos)
        else:
            return
        self.__info.text = str(len(self.__marker.params))

class MySceneCanvas2(scene.SceneCanvas):

    def __init__(self, parent):
        scene.SceneCanvas.__init__(self, parent=parent, app="wx", size=(10, 10))

        self.unfreeze()

        grid = self.central_widget.add_grid(margin=0)
        self.__view = grid.add_view(row=0, col=0)
        self.__view.camera = "panzoom"
        self.__view.camera.interactive = True
        self.__view.camera.rect = (0, 0, 1, 1)
        self.__markers = []

        self.__info = scene.Text(
            parent   = self.scene,
            pos      = (10, 10),
            anchor_x = "left",
            anchor_y = "bottom",
            color    = (1, 0, 0, 1),
            text     = "0"
        )
        
        self.freeze()

        self.show()

    def on_key_press(self, event):
        N = 100
        if event.text == "Q":
            self.__view.camera.rect = (0, 0, 1, 1)
        elif event.text == "A":
            for i in range(N):
                self.__markers.append(scene.Markers(
                    parent     = self.__view.scene,
                    pos        = np.random.rand(1, 2),
                    face_color = np.random.rand(1, 4)[0]
                ))
        elif event.text == "Z":
            for i in range(N):
                if len(self.__markers) == 0:
                    break
                random_marker = random.choice(self.__markers)
                random_marker.set_data(parent=None)
                self.__markers.remove(random_marker)
        else:
            return
        self.__info.text = str(len(self.__markers))

class MyFrame(wx.Frame):

    def __init__(self):
        wx.Frame.__init__(self, None)

        self.panel1 = wx.Panel(self)
        self.panel1.Bind(wx.EVT_SIZE, self.onSize)
        self.sc1 = MySceneCanvas1(self.panel1)

        self.panel2 = wx.Panel(self)
        self.panel2.Bind(wx.EVT_SIZE, self.onSize)
        self.sc2 = MySceneCanvas2(self.panel2)

        sizer = wx.BoxSizer(wx.HORIZONTAL)
        sizer.Add(self.panel1, 1, wx.EXPAND|wx.RIGHT, 5)
        sizer.Add(self.panel2, 1, wx.EXPAND)
        self.SetSizer(sizer)
    
    def onSize(self, event):
        self.sc1.size = self.panel1.GetSize()
        self.sc2.size = self.panel2.GetSize()
        event.Skip()

app = wx.App()
frame = MyFrame()
frame.Show()
app.MainLoop()
```

**Comment by brisvag** [comment]
Yes, you definitely want to use a single markers visual for this :) I'm not sure why it happens, but I noticed the performance hit of multiple visuals before, and it's quite a problem in some applications. We need to investigate!

In any case, I'm sure `MeshVisual` should be perfect doing what you need.

**Comment by sebae1** [comment]
> Yes, you definitely want to use a single markers visual for this :) I'm not sure why it happens, but I noticed the performance hit of multiple visuals before, and it's quite a problem in some applications. We need to investigate!
> 
> In any case, I'm sure `MeshVisual` should be perfect doing what you need.

Yeees, I got it!
Well, I didn't know PolygonVisual itself is a combination of MeshVisual and LineVisual. 
Thanks for you advice and I will attach my dirty example for the others.

Issue closed!

```
from vispy import scene, app
from vispy.geometry import PolygonData
import numpy as np

canvas = scene.SceneCanvas(keys='interactive')
canvas.size = 800, 600
canvas.show()

vb = canvas.central_widget.add_view()
vb.camera = "panzoom"
vb.camera.rect = (0, 0, 1, 1)

target_vertices = [
    # Every single line represents a polygon
    [[0.6, 0.8], [0.8, 0.8], [0.7, 0.7]], # a triangle
    [[0.1, 0.1], [0.3, 0.1], [0.4, 0.2], [0, 0.4]], # a square
    [[0.7, 0.1], [0.6, 0.2], [0.75, 0.3], [0.9, 0.2], [0.8, 0.1]] # a pentagon
]
target_face_colors = [
    # Face colors of the above polygons
    [1, 0, 0, 1],
    [0, 1, 0, 1],
    [0, 0, 1, 1]
]

# Set vertices and etc.
vertices = []
face_colors = []
faces = []
for i, v in enumerate(target_vertices):
    data = PolygonData(vertices=np.array(v))
    pts, tris = data.triangulate()
    faces.extend(list(tris.astype(np.uint32) + len(vertices)))
    vertices.extend(list(pts))
    for j in range(len(tris)):
        face_colors.append(target_face_colors[i])
vertices = np.array(vertices)
faces = np.array(faces)

scene.Mesh(
    parent      = vb.scene,
    vertices    = vertices,
    faces       = faces,
    face_colors = face_colors
)

if __name__ == "__main__":
    app.run()
```

**Comment by brisvag** [comment]
Sweet! If you still want the lines around it, you shoudl check out the [wireframe filter](https://github.com/vispy/vispy/blob/545abddf3cf5d7e672cb48fe8e68779fc338cc2d/vispy/visuals/filters/mesh.py#L619) for meshes :)

**Comment by sofroniewn** [comment]
In the napari shapes layer we have code for combining many polygons into a single mesh - including edges with nice bevels and joins which gives a nice performance boost. https://github.com/napari/napari/tree/main/napari/layers/shapes

It think it would make a lot of sense for us to contribute this code upstream into vispy into a new `PolygonsVisual`. It would reduce the amount of custom visual code in napari, enable everyone in the vispy community to benefit from it, and would be a great time to look into adding some performance optimizations (I think it will start to struggle above 300 shapes).

I wrote the napari code and would be more than happy to help guide the process, but I probably don't have bandwidth to do it myself.

If anyone is interested in this though please let me know!!

**Comment by sofroniewn** [comment]
The https://github.com/napari/napari/blob/main/napari/layers/shapes/_shape_list.py and https://github.com/napari/napari/blob/main/napari/layers/shapes/_mesh.py files are good places to start looking in
